framework/frequency_manager.hpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2022 Intel Corporation.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FREQUENCY_MANAGER_HPP
+#define FREQUENCY_MANAGER_HPP
+
+#define BASE_FREQ_PATH         "/sys/devices/system/cpu/cpu"
+#define SCALING_GOVERNOR       "/cpufreq/scaling_governor"
+#define SCALING_SETSPEED       "/cpufreq/scaling_setspeed"
+
+class FrequencyManager
+{
+private:
+#ifdef __linux__
+    int max_frequency_supported;
+    int min_frequency_supported;
+    std::vector<std::string> per_cpu_initial_scaling_governor;
+    std::vector<std::string> per_cpu_initial_scaling_setspeed;
+    int current_set_frequency;
+    std::vector<int> frequency_levels;
+    int frequency_level_idx = 0;
+    static constexpr int total_frequency_levels = 9;
+#endif
+
+    std::string read_file(const std::string &file_path)
+    {
+#ifdef __linux__
+        /* Read first line of given file */
+        char line[100]; //100 characters should be more than enough
+        FILE *file = fopen(file_path.c_str(), "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, file_path.c_str());
+            exit(EXIT_FAILURE);
+        }
+        fscanf(file, "%s", line);
+        fclose(file);
+        return std::string(line);
+#else
+        return "";
+#endif
+    }
+
+    int write_file(const std::string &file_path, const std::string &line)
+    {
+#ifdef __linux__
+        FILE *file = fopen(file_path.c_str(), "w");
+
+        if (file == NULL)
+            return 1;
+
+        fprintf(file, "%s", line.c_str());
+        fclose(file);
+#endif
+        return 0;
+    }
+
+    int get_frequency_from_file(const std::string &file_path)
+    {
+        /* Read frequency value from file */
+        int frequency;
+#ifdef __linux__
+        FILE *file = fopen(file_path.c_str(), "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, file_path.c_str());
+            exit(EXIT_FAILURE);
+        }
+        fscanf(file, "%d", &frequency);
+        fclose(file);
+#endif
+        return frequency;
+    }
+
+    void populate_frequency_levels()
+    {
+#ifdef __linux__
+        frequency_levels.push_back(max_frequency_supported);
+        frequency_levels.push_back(min_frequency_supported); 
+
+        std::vector<int> tmp = frequency_levels;
+
+        while (frequency_levels.size() != total_frequency_levels)
+        {
+            std::sort(tmp.begin(), tmp.end(), std::greater<int>());
+            for (int idx = 1; idx < tmp.size(); idx++)
+                frequency_levels.push_back((tmp[idx] + tmp[idx - 1]) / 2);
+            tmp = frequency_levels;
+        }
```

Comment:
Yes, i thought about this. the problem is in generating frequency values of all levels ping ponging in between, `3.6, 1.2, 3.6, 1.2, 2.4, 3.0, 1.8, 3.3, 2.7, 1.5`. This is the best algorithm i could come up in generating this all levels of frequency. I guess since n is 9, this should be fine.

Reply:
We'll worry about this later. Thanks.